scintellometry/io/arochime.py

```diff
@@ -166,6 +168,203 @@ def open(self, number=0):
             self.current_file_number = number
         return self.fh_raw
 
+class AROCHIMEVdifData(SequentialFile):
+
+    telescope = 'arochime-vdif'
+
+    def __init__(self, raw_files, blocksize, samplerate, fedge, fedge_at_top,
+                 time_offset=0.0*u.s, dtype='cu4bit,cu4bit', comm=None):
+        """ARO data acquired with a CHIME correlator, saved in VDIF format.
+        Files are 2**16 time, 2 pol, 1024 freq, at 800MHz / (2*1024) samplerate
+        Read with baseband VDIF package, need byte to sample conversions for folding
+        """
+
+        fh = vdif.open(raw_files[0], 'rs', sample_rate=samplerate)
+        self.time0 = fh.tell(unit='time')
+        self.npol = fh.nthread
+        nchan = fh.nchan
+        self.samplerate = samplerate
+        self.fedge_at_top = fedge_at_top
+        if fedge.isscalar:
+            self.fedge = fedge
+            f = fftshift(fftfreq(nchan, (2./samplerate).to(u.s).value)) * u.Hz
+            if fedge_at_top:
+                self.frequencies = fedge - (f-f[0])
+            else:
+                self.frequencies = fedge + (f-f[0])
+        else:
+            assert fedge.shape == (nchan,)
+            self.frequencies = fedge
+            if fedge_at_top:
+                self.fedge = self.frequencies.max()
+            else:
+                self.fedge = self.frequencies.min()
+
+        self.dtsample = (nchan * 2 / samplerate).to(u.s)
+        if comm is None or comm.rank == 0:
+            print("In AROCHIMEVdifData, calling super")
+            print("Start time: ", self.time0.iso)
+
+        super(AROCHIMEVdifData, self).__init__(raw_files, blocksize, dtype, nchan,
+                                           comm=comm)
+        if self.filesize % self.fh_raw.header0.framesize != 0:
+            raise ValueError("File size is not an integer number of packets")
+
+        self.filesize = (self.filesize // self.fh_raw.header0.framesize *
+                         self.fh_raw.header0.payloadsize)
+
+    def open(self, number=0):
+        """Open a new file in the sequence.
+
+        Parameters
+        ----------
+        file_number : int
+            The number of the file to open.  Default is 0, i.e., the first one.
+        """
+        if number != self.current_file_number:
+            self.close()
+            self.fh_raw = vdif.open(self.files[number], 'rs', sample_rate=(1/self.dtsample).to(u.Hz))
+            self.current_file_number = number
+            
+        return self.fh_raw
+
+    def seek_record_read(self, offset, count):
+        """Read count samples starting from offset (also in samples)"""
+        self.seek(offset)
+        return self.record_read(count)
+
+    def record_read(self, count):
+        return self.read(count).view('c8,c8')
```

Comment:
This is not quite general, e.g., if we were to read just one polarisation, or, if magically, data were unchannelized. Since we're only working around an annoying mistake in `fold.py`, how about just returning `self.read(count)` here and correcting the behaviour there (see other comment).